File: engine/src/theseus_engine/coverage.py
"""Coverage routes: visit ALL reachable floor, not one goal.

Boustrophedon ("as the ox plows") sweep: pick vertical lanes spaced
lane_width apart, walk each lane's reachable runs alternately up and
down, and hop between runs with A*. Use cases: patrol mode, "sweep the
house for my keys" (pair with the detector), vacuum-style coverage.

The guarantee is practical, not perfect: pockets narrower than the lane
spacing can slip between lanes. coverage_fraction() reports honestly
what fraction of reachable floor the route passes within lane_width of —
demos and tests assert on that number instead of pretending.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass

from . import astar
from .geometry import Cell
from .grid import OccupancyGrid, PlanParams


@dataclass
class CoverageResult:
    cells: list[Cell]        # the full walkable route, start included
    visit_points: list[Cell]  # sweep waypoints in visiting order
    lanes: int
    skipped: int             # sweep points no route could reach


def _reachable(grid: OccupancyGrid, start: Cell,
               params: PlanParams) -> set[Cell]:
    """Cells reachable from start under the canonical edge rules."""
    if not grid.in_bounds(start):
        return set()
    seen = {start}
    q = deque([start])
    while q:
        u = q.popleft()
        for v, _step in grid.neighbors8(u):
            if v not in seen and grid.edge_cost(u, v, params) < float("inf"):
                seen.add(v)
                q.append(v)
    return seen


def coverage_route(grid: OccupancyGrid, start: Cell, params: PlanParams,
                   lane_width: float = 0.6) -> CoverageResult | None:
    reach = _reachable(grid, start, params)
    if not reach:
        return None
    lane_step = max(1, round(lane_width / grid.cell_size))
    visit: list[Cell] = []
    lanes = 0
    upward = True
    for x in range(lane_step // 2, grid.width, lane_step):
        ys = sorted(y for (cx, y) in reach if cx == x)
        if not ys:
            continue
        lanes += 1
        # contiguous vertical runs (an obstacle splits a lane)
        runs: list[tuple[int, int]] = []
        y0 = prev = ys[0]
        for y in ys[1:]:
            if y == prev + 1:
                prev = y
            else:
                runs.append((y0, prev))
                y0 = prev = y
        runs.append((y0, prev))
        if not upward:
            runs.reverse()
        for lo, hi in runs:
            a, b = ((x, lo), (x, hi)) if upward else ((x, hi), (x, lo))
            visit.append(a)
            if b != a:
                visit.append(b)
        upward = not upward
    cells: list[Cell] = [start]
    cur = start
    skipped = 0
    for wp in visit:
        res = astar.plan(grid, cur, wp, params)
        if res is None:
            skipped += 1
            continue
        cells.extend(res.cells[1:])
        cur = wp
    return CoverageResult(cells, visit, lanes, skipped)


def coverage_fraction(grid: OccupancyGrid, route: list[Cell],
                      params: PlanParams, lane_width: float) -> float:
    """Fraction of floor reachable from the route's start that lies
    within lane_width (Chebyshev) of some route cell."""
    if not route:
        return 0.0
    reach = _reachable(grid, route[0], params)
    if not reach:
        return 0.0
    r = max(1, round(lane_width / grid.cell_size))
    covered: set[Cell] = set()
    for (x, y) in route:
        for dy in range(-r, r + 1):
            for dx in range(-r, r + 1):
                covered.add((x + dx, y + dy))
    hit = sum(1 for c in reach if c in covered)
    return hit / len(reach)

File: apps/navcore/Sources/NavCore/Coverage.swift
// Coverage routes — port of engine coverage.py. Boustrophedon sweep:
// vertical lanes spaced laneWidth apart, walk each lane's reachable runs
// alternately up and down, hop between runs with A*. The guarantee is
// practical, not perfect: coverageFraction reports honestly what
// fraction of reachable floor the route passes within laneWidth of.

public struct CoverageResult: Sendable {
    public var cells: [Cell]        // the full walkable route
    public var visitPoints: [Cell]  // sweep waypoints in visiting order
    public var lanes: Int
    public var skipped: Int         // sweep points no route could reach
}

/// Cells reachable from start under the canonical edge rules.
func reachable(grid: OccupancyGrid, start: Cell,
               params: PlanParams) -> Set<Cell> {
    if !grid.inBounds(start) {
        return []
    }
    var seen: Set<Cell> = [start]
    var queue = [start]
    var head = 0
    while head < queue.count {
        let u = queue[head]
        head += 1
        for (v, _) in grid.neighbors8(u) {
            if !seen.contains(v) && grid.edgeCost(u, v, params) < INF {
                seen.insert(v)
                queue.append(v)
            }
        }
    }
    return seen
}

public func coverageRoute(grid: OccupancyGrid, start: Cell,
                          params: PlanParams,
                          laneWidth: Double = 0.6) -> CoverageResult? {
    let reach = reachable(grid: grid, start: start, params: params)
    if reach.isEmpty {
        return nil
    }
    // Python round() — banker's, not .rounded()
    let laneStep = max(1, Int((laneWidth / grid.cellSize)
                              .rounded(.toNearestOrEven)))
    var visit: [Cell] = []
    var lanes = 0
    var upward = true
    for x in stride(from: laneStep / 2, to: grid.width, by: laneStep) {
        let ys = reach.filter { Int($0.x) == x }.map { Int($0.y) }.sorted()
        if ys.isEmpty {
            continue
        }
        lanes += 1
        // contiguous vertical runs (an obstacle splits a lane)
        var runs: [(Int, Int)] = []
        var y0 = ys[0]
        var prev = ys[0]
        for y in ys.dropFirst() {
            if y == prev + 1 {
                prev = y
            } else {
                runs.append((y0, prev))
                y0 = y
                prev = y
            }
        }
        runs.append((y0, prev))
        if !upward {
            runs.reverse()
        }
        for (lo, hi) in runs {
            let a = upward ? Cell(Int32(x), Int32(lo)) : Cell(Int32(x), Int32(hi))
            let b = upward ? Cell(Int32(x), Int32(hi)) : Cell(Int32(x), Int32(lo))
            visit.append(a)
            if b != a {
                visit.append(b)
            }
        }
        upward = !upward
    }
    var cells = [start]
    var cur = start
    var skipped = 0
    for wp in visit {
        guard let res = plan(grid: grid, start: cur, goal: wp,
                             params: params) else {
            skipped += 1
            continue
        }
        cells.append(contentsOf: res.cells.dropFirst())
        cur = wp
    }
    return CoverageResult(cells: cells, visitPoints: visit, lanes: lanes,
                          skipped: skipped)
}

/// Fraction of floor reachable from the route's start that lies within
/// laneWidth (Chebyshev) of some route cell.
public func coverageFraction(grid: OccupancyGrid, route: [Cell],
                             params: PlanParams,
                             laneWidth: Double) -> Double {
    if route.isEmpty {
        return 0.0
    }
    let reach = reachable(grid: grid, start: route[0], params: params)
    if reach.isEmpty {
        return 0.0
    }
    let r = max(1, Int((laneWidth / grid.cellSize)
                       .rounded(.toNearestOrEven)))
    var covered = Set<Cell>()
    for c in route {
        for dy in -r...r {
            for dx in -r...r {
                covered.insert(Cell(c.x + Int32(dx), c.y + Int32(dy)))
            }
        }
    }
    let hit = reach.reduce(0) { covered.contains($1) ? $0 + 1 : $0 }
    return Double(hit) / Double(reach.count)
}
